### PR TITLE
nixos/zigbee2mqtt: apply rfc42 and harden systemd unit

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

```
This is what is still exposed, and it allows me to control my lamps from
within home-assistant.

✗ PrivateNetwork=                                             Service has access to the host's network                                            0.5
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                               0.3
✗ DeviceAllow=                                                Service has a device ACL with some special devices                                  0.1
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                    0.2
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                       0.1
✗ MemoryDenyWriteExecute=                                     Service may create writable executable memory mappings                              0.1

→ Overall exposure level for zigbee2mqtt.service: 1.1 OK 🙂
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
The nixos test now fails because the serial port is required to be mounted into the /dev mount namespace. This means we cannot test this like that anymore.

**Comment by sweber83** [comment]
@mweinelt 
I had a go at testing your PR.
I'm having problems accessing the serial device.
Here's my log output from zigbee2mqtt:
```
Apr 29 20:49:50 homepi systemd[1]: Starting Zigbee2mqtt Service...
Apr 29 20:49:50 homepi systemd[1]: Started Zigbee2mqtt Service.
Apr 29 20:49:52 homepi zigbee2mqtt[923316]: Zigbee2MQTT:debug 2021-04-29 20:49:52: Loaded state from file /var/lib/zigbee2mqtt/state.json
Apr 29 20:49:52 homepi zigbee2mqtt[923316]: Zigbee2MQTT:info  2021-04-29 20:49:52: Logging to console and directory: '/var/lib/zigbee2mqtt/log/2021-04-29.20-49-51' filename: log.txt
Apr 29 20:49:52 homepi zigbee2mqtt[923316]: Zigbee2MQTT:debug 2021-04-29 20:49:52: Removing old log directory '/var/lib/zigbee2mqtt/log/2021-04-29.20-49-16'
Apr 29 20:49:52 homepi zigbee2mqtt[923316]: Zigbee2MQTT:info  2021-04-29 20:49:52: Starting Zigbee2MQTT version 1.18.1 (commit #unknown)
Apr 29 20:49:52 homepi zigbee2mqtt[923316]: Zigbee2MQTT:info  2021-04-29 20:49:52: Starting zigbee-herdsman (0.13.71)
Apr 29 20:49:52 homepi zigbee2mqtt[923316]: Zigbee2MQTT:debug 2021-04-29 20:49:52: Using zigbee-herdsman with settings: '{"adapter":{"concurrent":null,"delay":null},"backupPath":"/var/lib/zigbee2mqtt/coordinator_backup.json","databaseBackupPath":"/var/lib/zigbee2mqtt/database.db.backup","databasePath":"/var/lib/zigbee2mqtt/database.db","network":{"channelList":[25],"extendedPanID":[221,221,221,221,221,221,221,221],"networkKey":"HIDDEN","panID":6754},"serialPort":{"path":"/dev/serial/by-id/usb-Texas_Instruments_CC2538_USB_CDC-if00"}}'
Apr 29 20:49:53 homepi zigbee2mqtt[923316]: Zigbee2MQTT:error 2021-04-29 20:49:53: Error while starting zigbee-herdsman
Apr 29 20:49:53 homepi zigbee2mqtt[923316]: Zigbee2MQTT:error 2021-04-29 20:49:53: Failed to start zigbee
Apr 29 20:49:53 homepi zigbee2mqtt[923316]: Zigbee2MQTT:error 2021-04-29 20:49:53: Exiting...
Apr 29 20:49:53 homepi zigbee2mqtt[923316]: Zigbee2MQTT:error 2021-04-29 20:49:53: Error: Error while opening serialport 'Error: Error: Permission denied, cannot open /dev/serial/by-id/usb-Texas_Instruments_CC2538_USB_CDC-if00'
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at Znp.<anonymous> (/nix/store/m4bg58alnnmdaj1fmr5qjhv96clv994i-node_zigbee2mqtt-1.18.1/lib/node_modules/zigbee2mqtt/node_modules/zigbee-herdsman/dist/adapter/z-stack/znp/znp.js:136:32)
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at Generator.next (<anonymous>)
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at /nix/store/m4bg58alnnmdaj1fmr5qjhv96clv994i-node_zigbee2mqtt-1.18.1/lib/node_modules/zigbee2mqtt/node_modules/zigbee-herdsman/dist/adapter/z-stack/znp/znp.js:27:71
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at new Promise (<anonymous>)
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at __awaiter (/nix/store/m4bg58alnnmdaj1fmr5qjhv96clv994i-node_zigbee2mqtt-1.18.1/lib/node_modules/zigbee2mqtt/node_modules/zigbee-herdsman/dist/adapter/z-stack/znp/znp.js:23:12)
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at SerialPort.<anonymous> (/nix/store/m4bg58alnnmdaj1fmr5qjhv96clv994i-node_zigbee2mqtt-1.18.1/lib/node_modules/zigbee2mqtt/node_modules/zigbee-herdsman/dist/adapter/z-stack/znp/znp.js:134:49)
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at SerialPort._error (/nix/store/m4bg58alnnmdaj1fmr5qjhv96clv994i-node_zigbee2mqtt-1.18.1/lib/node_modules/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/stream/lib/index.js:198:14)
Apr 29 20:49:53 homepi zigbee2mqtt[923316]:     at /nix/store/m4bg58alnnmdaj1fmr5qjhv96clv994i-node_zigbee2mqtt-1.18.1/lib/node_modules/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/stream/lib/index.js:242:1
```
And here my systemd unit:
```
# /nix/store/chb29bi7q1wn569gfabkmcgq1am7y8nz-unit-zigbee2mqtt.service/zigbee2mqtt.service
[Unit]
After=network.target
Description=Zigbee2mqtt Service

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/xq7mbznjqgjcjqgwgg9kh85igaw9gg2g-glibc-locales-2.32-39/lib/locale/locale-archive"
Environment="PATH=/nix/store/xs02fpnpkqfrhqqfsxx3lpj48wrapd00-coreutils-8.32/bin:/nix/store/jl2zrqjj1dzi9r9in7pl19zh47vsypn2-findutils-4.7.0/bin:/nix/store/mn3g1jdw67ibv2b7h887z9l4kh31sk29-gnugrep-3.6/bin:/nix/store/wwjpgh4yzw34b2k5lm7cqj564ccbr4d3-gnused-4.8/bin:/nix/store/ili7vamzgwwsndmq7rx6bn0ig8cr3wdw-systemd-247.3/bin:/nix/store/xs02fpnpkqfrhqqfsxx3lpj48wrapd00-coreutils-8.32/sbin:/nix/store/jl2zrqjj1dzi9r9in7pl19zh47vsypn2-findutils-4.7.0/sbin:/nix/store/mn3g1jdw67ibv2b7h887z9l4kh31sk29-gnugrep-3.6/sbin:/nix/store/wwjpgh4yzw34b2k5lm7cqj564ccbr4d3-gnused-4.8/sbin:/nix/store/ili7vamzgwwsndmq7rx6bn0ig8cr3wdw-systemd-247.3/sbin"
Environment="TZDIR=/nix/store/jq39whk68mj29wvzn2875ja019z8bp9g-tzdata-2020f/share/zoneinfo"
Environment="ZIGBEE2MQTT_DATA=/var/lib/zigbee2mqtt"



BindPaths=/dev/serial/by-id/usb-Texas_Instruments_CC2538_USB_CDC-if00
CapabilityBoundingSet=
DeviceAllow=/dev/serial/by-id/usb-Texas_Instruments_CC2538_USB_CDC-if00
DevicePolicy=closed
ExecStart=/nix/store/m4bg58alnnmdaj1fmr5qjhv96clv994i-node_zigbee2mqtt-1.18.1/bin/zigbee2mqtt
ExecStartPre=/nix/store/m1r33z9p7b8fbrhwjnz0ppbc6i44v65h-unit-script-zigbee2mqtt-pre-start/bin/zigbee2mqtt-pre-start
LockPersonality=true
MemoryDenyWriteExecute=false
NoNewPrivileges=true
PrivateDevices=true
PrivateTmp=true
PrivateUsers=true
ProcSubset=pid
ProtectClock=true
ProtectControlGroups=true
ProtectHome=true
ProtectHostname=true
ProtectKernelLogs=true
ProtectKernelModules=true
ProtectKernelTunables=true
ProtectProc=invisible
ProtectSystem=strict
ReadWritePaths=/var/lib/zigbee2mqtt
RemoveIPC=true
Restart=on-failure
RestrictAddressFamilies=AF_INET
RestrictAddressFamilies=AF_INET6
RestrictNamespaces=true
RestrictRealtime=true
RestrictSUIDSGID=true
SystemCallArchitectures=native
SystemCallFilter=@system-service
SystemCallFilter=~@privileged
SystemCallFilter=~@resources
UMask=0077
User=zigbee2mqtt
WorkingDirectory=/var/lib/zigbee2mqtt
```
One thing, important to know, is that I did not build the whole machine with nixpkgs@59798a8c3f34b19c061ebdb5d407cecc5ddf3b0d but only imported the module from there.
The machine itself currently runs with nixpkgs@c6bd90d48fbb60297c9406edc1693b7aa66cd33b.
The reason for doing it this way, was that home-assistant did not build @59798a8c3f34b19c061ebdb5d407cecc5ddf3b0d.
I'm not sure if this could be a factor for the failure.

**Comment by mweinelt** [comment]
Try and disable PrivateUser, it might interfere with the assigned dialout group. I might move the group into SupplementaryGroups.

**Comment by sweber83** [comment]
I tried disabling `PrivateUsers`, but the issue persists.

Regarding the nixos test. Would it not be possible, to modify the test, so the parts of the hardening, which interfere, are disabled?

**Comment by mweinelt** [comment]
For some reason it cannot access `/dev/serial` at all. I've tried it with `/dev/ttyUSB0` and that belonged to to `root:dialout`, so no problem there.

```
total 0
drwxr-xr-x 8 root root  400 Apr 30 00:44 .
drwxr-xr-x 1 root root  114 May 20  2020 ..
drwxr-xr-x 2 root root  180 Apr 30 00:44 char
lrwxrwxrwx 1 root root   13 Apr 30 00:44 fd -> /proc/self/fd
crw-rw-rw- 1 root root 1, 7 Apr 30 00:44 full
drwxr-xr-x 2 root root    0 Jan  1  1980 hugepages
lrwxrwxrwx 1 root root   28 Apr 30 00:44 log -> /run/systemd/journal/dev-log
drwxrwxrwt 2 root root   40 Jan  1  1970 mqueue
crw-rw-rw- 1 root root 1, 3 Apr 30 00:44 null
crw-rw-rw- 1 root root 5, 2 Apr 30 00:44 ptmx
drwxr-xr-x 2 root root    0 Jan  1  1970 pts
crw-rw-rw- 1 root root 1, 8 Apr 30 00:44 random
drwx------ 3 root root   60 Apr 30 00:44 serial
drwxrwxrwt 2 root root  160 Apr 29 21:46 shm
lrwxrwxrwx 1 root root   15 Apr 30 00:44 stderr -> /proc/self/fd/2
lrwxrwxrwx 1 root root   15 Apr 30 00:44 stdin -> /proc/self/fd/0
lrwxrwxrwx 1 root root   15 Apr 30 00:44 stdout -> /proc/self/fd/1
crw-rw-rw- 1 root root 5, 0 Apr 30 00:44 tty
crw-rw-rw- 1 root root 1, 9 Apr 30 00:44 urandom
crw-rw-rw- 1 root root 1, 5 Apr 30 00:44 zero
ls: cannot open directory '/dev/serial': Permission denied
ls: cannot access '/dev/serial/by-id/': Permission denied
```

**Comment by mweinelt** [comment]
@sweber83 Updated. Please, please retest. I want to get this into 21.05. I've also converted the module to RFC 0042, so `config` has been renamed to `settings`, the attrset in there can stay as is. Release notes have been added accordingly.

File: nixos/modules/services/misc/zigbee2mqtt.nix
Comment (sweber83) [inline nixos/modules/services/misc/zigbee2mqtt.nix:26]: @mweinelt 
Did you intentionally leave out the devices.yaml-configuration?
When I leave that out, zigbee2mqtt puts the device config in configuration.yaml, which is overwritten every start.

Apart from that, your PR works fine now.

Comment (mweinelt) [reply nixos/modules/services/misc/zigbee2mqtt.nix:26]: Odd, I wrote a comment in the new preset to that regard, not sure where it went missing. Will check up asap.

Comment (mweinelt) [reply nixos/modules/services/misc/zigbee2mqtt.nix:26]: Fixed.

Comment (sweber83) [reply nixos/modules/services/misc/zigbee2mqtt.nix:26]: @mweinelt 
Works fine, now :+1: 
Really nice that the test continues to work.

How did you determine the systemd settings for hardening?
Did you use some kind of tool, or did you work it out manually?

**Comment by mweinelt** [comment]
I'm working on an assertion, that pokes users to migrate to settings still, then this should be good to merge.

**Comment by mweinelt** [comment]
Finally, adding myself to maintainers, given the huge mess I've created!

@sweber83 I'll merge as soon as you hit approve.

**Review by sweber83 — APPROVED** [review]
@mweinelt 
Really nice improvements :+1:

**Pull request merged**
